Initial tracing support for functions (referred to as `tracee functions`)

mhong opened the pull request:
with the following signature: `(State, Data) -> (State, Result)`

This is designed to support the computation of a training step, where the
function takes and returns a Model (aka "state"), containing a collection of
trainable parameters. The function also takes as input a Data batch.

The additional output of `Result` can be used for the training step to produce
metrics (e.g. loss value).

Commit:
Initial tracing support for functions (referred to as `tracee functions`) with
the following signature: `(State, Data) -> (State, Result)`

This is designed to support the computation of a training step, where the
function takes and returns a Model (aka "state"), containing a collection of
trainable parameters. The function also takes as input a Data batch.

The additional output of `Result` can be used for the training step to produce
metrics (e.g. loss value).

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Also cc'ing @jekbradbury as FYI

pschuh commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+  let graph = TF_NewGraph()
+
+  // Used to create unique trace graph node names and graph function names.
+  var traceGraphObjectCounter = 0
```
I don't see this used to uniquify node names. Would you not need this entirely if you used append_hash_to_fn_name below?

mhong replied:
I moved this to `_RuntimeConfig` so that it lives across the lifetimes of `TraceContext` instances. Also renamed it to `traceGraphFunctionCounter`. 

We don't need a "trace graph node counter" at this swift level (any more), since that's done at the C API impl of `TFE_AddEagerOpToGraph()`.

pschuh commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  }
+}
+
+// TODO: assess if this protocol should be foled into TensorArrayProtocol.
```
foled -> folded ?

mhong replied:
Fixed.

Commit:
Addressed feedback.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+  case tracing(TraceContext)
+
+  // Return nil if we are not in tracing mode.
+  var getFunction : TraceContext? {
```
* Avoid using `get` with an object in Swift names.
* Why is this called "function" instead of "context"? "Context" seems to be clearer.
* Remove the space before the colon after variable names.
```suggestion
  var context: TraceContext? {
```

bgogul replied:
+1, should this be context?

mhong replied:
Done.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+
+  // Return nil if we are not in tracing mode.
+  var getFunction : TraceContext? {
+    switch self {
```
When there are two cases and you only want to match one case, `switch` is not necessary. You can write:
```swift
guard case let .tracing(trace) else { return nil }
return trace
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+                                     outputs: outputs)
+}
+
+public extension TensorArrayProtocol {
```
This is not used outside this module. Make private.
```suggestion
private extension TensorArrayProtocol {
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+
+public extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  func _getCTensorHandles() -> [CTensorHandle] {
```
Avoid using "get" and use a computed property when the method does not take any arguments.
```suggestion
  var cTensorHandles: [CTensorHandle] {
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: Int(self._tensorHandleCount))
+    debugLog("Unpacking handles into buffer.")
+    self._unpackTensorHandles(into: buffer)
```
```suggestion
    _unpackTensorHandles(into: buffer)
```

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+public extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  func _getCTensorHandles() -> [CTensorHandle] {
+    debugLog("Getting \(self._tensorHandleCount) C handles.")
```
Our local convention is to not use `self.` unless it's required.

```suggestion
    debugLog("Getting \(_tensorHandleCount) C handles.")
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+               \(maxReturnValueCount) return values.
+               """)
+    var returnValues: [CTensorHandle?]
+    returnValues = Array(repeating: nil, count: maxReturnValueCount)
```
```swift
var returnValues = [CTensorHandle?](repeating: nil, count: maxReturnValueCount)
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    let status = TF_NewStatus()
+    debugLog("Copying buffer content to output handles.")
+    var output: [CTensorHandle] = []
+    for idx in 0..<self._tensorHandleCount {
```
* Nit: We normally just use `i` (or `index` if spelled out) in the code base, but I don't have a strong opinion.
* It'd be slightly better if you convert `_tensorHandleCount` to `Int` in the range, instead of converting inside the loop.

```suggestion
    for i in 0..<Int(_tensorHandleCount) {
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  }
+}
+
+public extension TensorGroup {
```
Can this be `private`?

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: input.count)
+    let status = TF_NewStatus()
+    // copy input to buffer
+    for (idx, inputTensorHandle) in input.enumerated() {
```
`i`, if you care, but i don't have a strong opinion!
```suggestion
    for (i, inputTensorHandle) in input.enumerated() {
```

mhong replied:
Changed to `i`.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
+    assert(Self._tensorHandleCount == input.count)
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: input.count)
```
Fit within 80 columns

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  // Create an instance based on `inputs`, which can be symbolic (e.g. when
+  // creating a symbolic input to tracee) or concrete (e.g. when creating a
+  // final output of executing the tracee).
+  func createInstance(_owning inputs: [CTensorHandle]) -> Self
```
* It's fine to add this to `TensorArrayProtocol` today (and remove `TensorArrayProtocolEnhanced`). Otherwise generic algorithms won't be compatible with the `trace(with:in:)` operator. `TensorArrayProtocol` is the public API that users interact with.
* Does this rely on `self`, i.e. do `inputs` have to match the shapes of each tensor in `self`?
  * If so, it's called a factory method in Swift, which could be called `_makeInstance(owning:)` according to the naming convention.
  * It it does not rely on `self`, this should be a static. Since the semantics of this method is initialization, it should an initializer requirement on `TensorArrayProtocol`, i.e. `init(_owning:)`.

Depending on whether it becomes an initializer or a factory method. In the initializer case, the first argument label should be underscored. In the factory method case, the method name should be underscored but the argument label need not be.

mhong replied:
Thanks for the style guidance. This is a factory method, and I named it accordingly.

Merging the protocols SGTM. I'm deferring that work (left a TODO), because it requires we implement this new method `_makeInstance` for each struct/class that conforms to `TensorArrayProtocol` (e.g. `Tensor._makeInstance()`).

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    let resultTensors: [CTensorHandle] = Array(returnValues.dropFirst(
+                                                 Int(state._tensorHandleCount)))
+    let result = Result(_copying: resultTensors)
+    return(newState, result)
```
```suggestion
    return (newState, result)
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    debugLog("Creating output model instance.")
+    let newState = state.createInstance(_owning: Array(returnValues.dropLast(
+                                                         Data._typeList.count)))
+    let resultTensors: [CTensorHandle] = Array(returnValues.dropFirst(
```
```suggestion
    let resultTensors: [CTensorHandle] = 
      Array(returnValues.dropFirst(Int(state._tensorHandleCount)))
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    debugLog("Getting input state tensor handles.")
+    let inputStateTensorHandles =  oldState._getCTensorHandles()
+    var inputTensors = inputStateTensorHandles.map {
+      Tensor<Float>(handle: TensorHandle(_owning: $0)) }
```
```suggestion
      Tensor<Float>(handle: TensorHandle(_owning: $0))
    }
```

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    fatalError("Not in tracing mode!.")
+  }
+  _RuntimeConfig.traceState = .notTracing
+
```
Remove extra line?

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+    debugLog("Getting input state tensor handles.")
+    let inputStateTensorHandles =  oldState._getCTensorHandles()
+    var inputTensors = inputStateTensorHandles.map {
+      Tensor<Float>(handle: TensorHandle(_owning: $0)) }
```
Why is it specialized on `Float`?

mhong replied:
This is a short-cut we took for this initial PR. Left a `// TODO: support other dtypes.` above, and I repeated that TODO here.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+    var traceGraphOutputs: [CTensorHandle] = []
+    // Points to an element in `returnValues`.
+    var returnValueIdx = 0
+    for output in outputs {
```
Use `outputs.enumerated()` to get both an index and a value, instead of counting manually.

mhong replied:
In the next PR we'll extend the code here that requires manual counting. Left this comment block on the code here.

```
    // We manually increment `returnValueIdx` below instead of using
    // `outputs.enumerated()`, because the logic will be extended in a future PR
    // that requires manual counting.
```

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+      debugLog("The traced function is:\n\(String(cString: funcDebugStr))")
+      free(funcDebugStr)
+    }
+
```
Remove extra line?

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  }
+  internalConsistencyCheck(inputSymbolicTensors.count == inputValueCount)
+  let symbolicState = state.createInstance(
+    _owning: Array(inputSymbolicTensors.dropLast(Data._typeList.count)))
```
It's better if you make this method (`TensorArrayProtocol._makeInstance(_owning:)` or `TensorArrayProtocol.init(_owning:)`) take a generic type `C : Collection where C.Element == CTensorHandle`. Otherwise it's inefficient (you had to create a new `Array` every time) and complicates callers.

mhong replied:
Nice suggestion! Done.

bgogul commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+  //
+  // TODO: If some tensors in `x` are not used within `foo()`, they can be
+  // pruned away in the inputs to the trace graph function.
+  var symbolicInputs: [TF_Output] = []
```
nit: doesn't matter for semantics, but shoudl this be TF_Input?

mhong replied:
Using TF_Output is the convention (e.g. also in TFLowerGraph.cpp). TF_Input appears to be deprecated.

bgogul commented on test/TensorFlowRuntime/tracer.swift:
```diff
@@ -0,0 +1,52 @@
+  let state = Tensor<Float>(2.0)
+  let data = Tensor<Float>(3.0)
+  let tracedFn = trace(with: state, in: tracee)
+  let (newState, result) = tracedFn(state, data)
```
Add a second call with different inputs?

mhong replied:
Nice suggestion! Done.

bgogul commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+  }
+
+  deinit {
+    TFE_DeleteTraceContext(cTraceContext)
```
nit, does the order of destruction matter? TraceContext seems to hold onto graph?

mhong replied:
Trace graph is created first, and then fed into a trace context object (which does not own the graph). So the destruction should be in the reverse ordering.

Commit:
Addressed more feedback.

Commit:
Addressed more feedback and enhanced unit testing to run the created trace graph
function multiple times.

This requires that we finalize the trace graph function *before* running the
return closure of `trace()`, so that in that return closure, we are no longer in
tracing mode, and focus on executing the trace graph function.

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
   }
 }
 
+// TraceContext contains the state as needed to build a trace graph function
```
Done

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Thanks for the comments and suggestions. I made a pass. PTAL.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,196 @@ public enum _ExecutionMode : Equatable {
+/// `addEagerOpToGraph()`). When the trace is finalized (via `finalize()`), the
+/// trace graph function can then be executed (via `execute()`) by the eager
+/// runtime.
+class TraceContext {
```
The entire `TraceContext` type can also be `private` because it's not in use elsewhere.

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -87,10 +87,196 @@ public enum _ExecutionMode : Equatable {
+      checkOk(status)
+      symbolicInputs.append(TF_Output(oper: result, index: 0))
+    }
+
```
Nit: Remove extra line?

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+    where C.Element == CTensorHandle
+}
+
+public func trace<State : TensorArrayProtocolEnhanced,
```
@jekbradbury and I discussed earlier and we'd like to name this API `graph(with:in:)`, instead of baking the technically inaccurate notion "trace" into the user's mental model.

If you think this feature is not ready yet, it's also okay to add an underscore to it to make it not show up in code completion as a supported feature.

mhong replied:
Done.

rxwei reviewed (approved):
Excellent!

Commit:
Addressed more feedback.

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+                                            outputs: outputTensorHandles)
+
+    debugLog("Creating output model instance.")
+    let newState = state._makeInstance(owning: Array(returnValues.dropLast(
```
Great point. Done.

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+  }
+  internalConsistencyCheck(inputSymbolicTensors.count == inputValueCount)
+  let symbolicState = state._makeInstance(
+    owning: Array(inputSymbolicTensors.dropLast(Data._typeList.count)))
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+
+private extension TensorGroup {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+  let symbolicState = state._makeInstance(
+    owning: Array(inputSymbolicTensors.dropLast(Data._typeList.count)))
+  let symbolicData = Data(
+    _copying: Array(inputSymbolicTensors.dropFirst(
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+                                                       Data._typeList.count)))
+    let resultTensors: [CTensorHandle] =
+      Array(returnValues.dropFirst(Int(state._tensorHandleCount)))
+    let result = Result(_copying: resultTensors)
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+                 Copying the \(i)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete).
+                 """)
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
```
I tried this, but then the use of `\(newHandle)` in `debugLog()` would trigger warnings that this is still an optional value. Since we want to print the unwrapped pointer value, I did not make this change.

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+  in fn: (State, Data) -> (State, Result)
+) -> (State, Data) -> (State, Result) {
+  debugLog("""
+             Tracing over a function with \(state._tensorHandleCount) input \
```
The indentation is set by my swift emacs mode, and I did not change it.

rxwei replied:
I see.  This is actually not about code style, but about the indentation of the string itself.  Swift multi-line string literals's indentation starts counting from the beginning of `"""`, which means that the this log string will be 2-space indented.

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+// Elements in `outputs` can come from two sources:
+// a) Symbolic tensors produced by tensor ops, and added as trace graph nodes.
+// b) Concrete tensors produced by host code (e.g. Tensor(1.0)).
+func finalizeTraceFn(fnName: String,
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+// TODO: Fold this protocol into TensorArrayProtocol.
+// This requires that we move concrete implementation such as
+// Tensor._makeInstance() to TensorGroup.swift.
+public protocol TensorArrayProtocolEnhanced : TensorArrayProtocol {
```
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+    debugLog("Getting input data tensor handles.")
+    let inputDataTensorHandles =  data.cTensorHandles
+    inputTensors.append(contentsOf: inputDataTensorHandles.map {
+                          Tensor<Float>(handle: TensorHandle(_owning: $0)) })
```
Done

Commit:
Some more style tweaks.

mhong commented:
@rxwei, on the issue of indentation of a multi-line debug string, such as:
```swift
debugLog("""
             Tracing over a function with \(state._tensorHandleCount) input \
...
```

The output appears properly indented:

```
stdout>>> [/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:670] Tracing over a function with 1 input state tensors and 1 input data tensors.
```

rxwei commented:
Ah ok! I misremembered it as being aligned with the open `"""`, but it's actually aligned with the closing `"""`.

mhong commented:
Thanks all for the review! I'm merging this PR to send out a follow-up one. Happy to address additional comments if any.
